### PR TITLE
Cache

### PR DESCRIPTION
**Pull request by SiaoChi**
add jest-test for user-module

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -32,4 +32,9 @@ lerna-debug.log*
 !.vscode/settings.json
 !.vscode/tasks.json
 !.vscode/launch.json
-!.vscode/extensions.json
\ No newline at end of file
+!.vscode/extensions.json
+
+# .env
```
Comment (RamoTsaiTW) [inline .gitignore]: Remove it

File: .gitignore [hunk 2]
```diff
@@ -32,4 +32,9 @@ lerna-debug.log*
...
+.env
+
+# other
```
Comment (RamoTsaiTW) [inline .gitignore]: remove

File: src/app.module.ts [hunk 3]
```diff
@@ -1,10 +1,40 @@
...
+    CacheModule.register({
+      isGlobal: true,
+      // this '@ts-ignore' is required for redis v4 config, due to the original version not support v4
+      // @ts-ignore
+      store: async () =>
+        await redisStore({
+          // after redis upgrade to v4 the property need to be set in socket object
+          socket: {
+            host: 'localhost',
+            port: 6379,
```
Comment (RamoTsaiTW) [inline src/app.module.ts]: Add the comment to remember using the environment variables to set the host and port.

File: src/users/users.service.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  UserProfileDto,
+} from '../dto/users-dto';
+import { Login } from '../model/users/login.entity';
+import { User } from '../model/users/user.entity';
+import { In, Repository } from 'typeorm';
+import * as bcrypt from 'bcrypt';
+import { AuthService } from '../utils/token.service';
+import { S3Service } from '../utils/s3.service';
+import { Tag } from '../model/users/tag.entity';
+// import { RedisService } from '../redis/redis.service';
```
Comment (RamoTsaiTW) [inline src/users/users.service.ts]: remove

File: src/users/users.service.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    const userInfo = await this.userRepository.findOne({
+      where: { user_id },
+    });
+    if (!userInfo)
+      throw new HttpException('user id not found', HttpStatus.BAD_REQUEST);
+
+    /* there are three scenarios for file
+    1. user upload file 
+    2. user choose default pictureURL
+    3. user didn't change any picture, it would only upload the original URL from database
```
Comment (RamoTsaiTW) [inline src/users/users.service.ts:124]: Good!

File: src/users/users.service.ts [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    userInfo.curr_problem = curr_problem;
+    const userInfoSave = await this.userRepository.save(userInfo);
+    await this.saveUserTags(userInfo.user_id, tags);
+    const expiredTime = 60 * 60 * 24 * 30;
+    await this.cacheManager.set(
+      `userId_profile_${user_id}`,
+      JSON.stringify(userInfoSave),
+      expiredTime,
+    );
+    await this.cacheManager.get(`userId_profile_${user_id}`);
```
Comment (RamoTsaiTW) [inline src/users/users.service.ts]: remove it

File: src/users/users.service.ts [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+      where: { user_id },
+      relations: ['tags'],
+    });
+    if (!userInfo)
+      throw new HttpException('user id not found', HttpStatus.BAD_REQUEST);
+    console.log('資料庫找id info');
+    return userInfo;
+  }
+
+  async updateUserProfile(body: UserProfileDto, file: object): Promise<object> {
```
Comment (RamoTsaiTW) [inline src/users/users.service.ts:98]: Please add the unit test for this function

File: .gitignore (hunk 1 above)
Comment (SiaoChi) [reply .gitignore]: done!

File: .gitignore (hunk 2 above)
Comment (SiaoChi) [reply .gitignore]: done!

File: src/app.module.ts (hunk 3 above)
Comment (SiaoChi) [reply src/app.module.ts]: done -> process.env.ENV

File: src/users/users.service.ts (hunk 4 above)
Comment (SiaoChi) [reply src/users/users.service.ts]: done

File: src/users/users.service.ts (hunk 6 above)
Comment (SiaoChi) [reply src/users/users.service.ts]: done

File: src/users/users.service.ts (hunk 7 above)
Comment (SiaoChi) [reply src/users/users.service.ts:98]: will do it later

**Review by RamoTsaiTW — APPROVED** [review]
Good job!!! :D

**Pull request merged**